File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.java
class Solution {
    
    private int solve(int i, String s, String[] dictionary, Integer[] dp){
        if(i >= s.length()) return 0;
        
        if(dp[i] != null) return dp[i];
        
        int leave = 1 + solve(i + 1, s, dictionary, dp);
        
        int take = s.length();
        for(String word : dictionary){
            if(i + word.length() > s.length()) continue;
            
            if(s.substring(i, i + word.length()).equals(word))
                take = Math.min(take, solve(i + word.length(), s, dictionary, dp));            
        }
        
        return dp[i] = Math.min(leave, take);
    }
    
    public int minExtraChar(String s, String[] dictionary) {
        Integer[] dp = new Integer[s.length()];
        
        int ans = solve(0, s, dictionary, dp);
        for(var i : dp)
            System.out.print(i + " ");
        System.out.println();
        return ans;
    }
}

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    int minExtraChar(string s, vector<string>& str) {
        unordered_set<string> set;
        for (auto m : str)
            set.insert(m);

        int n = s.size();
        vector<int> ans(n);

        for (int i = n - 1; i >= 0; i--) {
            int min = INT_MAX;
            for (int j = i; j < n; j++) {
                string curr = s.substr(i, j - i + 1);
                if (set.count(curr))
                    min = std::min(min, (j + 1 < n) ? ans[j + 1] : 0);
                else
                    min = std::min(min, ((j + 1 < n) ? ans[j + 1] : 0) + (j - i + 1));
            }

            ans[i] = min;
        }
        return ans[0];
    }
};